packages/client/src/types.ts

```diff
@@ -36,6 +36,7 @@ export interface WorkflowExecutionInfo {
   taskQueue: string;
   status: { code: proto.temporal.api.enums.v1.WorkflowExecutionStatus; name: WorkflowExecutionStatusName };
   historyLength: number;
+  historySizeBytes: number;
```

Comment:
I believe this prop (and some others) will not be available in some situations, notably on info returned from visibility APIs (ie. list workflows) for workflows that haven't yet completed execution (there might be other cases, I'm not sure about the exact extent of that limitation, and I assume it may possibly vary depending on which visibility backend is used).

We should therefore make that prop optionnal, and it would be pertinent to add some comments to that effect.

Also, we should have at least one test validating this prop through workflow client.

Reply:
Good catch @mjameswh

Reply:
One thing that I found confusing is that this is a different "historySizeBytes" that is coming from WorkflowExecutionInfo and not from WorkflowTaskStarted. This is not an optional field, that it defaults to 0, why would it not be available?

Reply:
@antlai-temporal If you dont know about the Visibility API and Visibility Store yet, get a look [here](https://docs.temporal.io/clusters#visibility) and [here](https://docs.temporal.io/cluster-deployment-guide#visibility-store).

The visibility store is optimized for efficient listing/searching across a large number of workflows. For that reason, it is eventual consistent, and not all details of a workflow are stored there (in particular, details that would be too large, that would change too often, that would have little value in search/list use cases, and other reasons).

For example, look at the `History Length` and `History Size` columns in the following picture:

<img width="2384" alt="Screenshot 2023-08-25 at 3 54 26 PM" src="https://github.com/temporalio/sdk-typescript/assets/6412169/b0b19b8e-e95b-49e5-bf1d-bc414846be52">

Reply:
Thanks James, but even if the value is missing in the database, we are directly reading from a proto that has non-optional int64 fields, so they should default to zero, right? I guess that is the reason we did not see it with the historyLength. 
I was wondering why the GUI did not show zeros, maybe they don't use protobufs...

Reply:
Now, back to our original discussion... The problem with reporting history length and history size (and probably other fields too) as 0 is simply that these values are surprising for developers who are not aware of the fact that the visibility API only report these values on workflow executions that are no longer executing. We periodically get user questions on that subject.

A solution to this can be as simple as adding doc comments on fields in that interface (I think pretty much all of it can be copied from elsewhere), then add extra notes on these specific fields saying that these may not be available in some contexts. However, I don't like the idea of saying something similar to "this field may reports 0 when doing visibily api and if the workflow is still executing"; marking that field as optional seems like a more appropriate thing to do (and then, mention that "that information may not be available when ...".

Reply:
The fact that we can explicitly set it to zero, and after de-serialization will ended up as missing is very confusing, I will do the change to explicitly set it to zero, and put a comment that zero means that the value may, or may not, have been set. 
I don't think we should talk about the visibility API in the SDK comments, the UI could show the missing fields as "pending", or add a comment pop-up. In any case, most users of the UI are not going to look at the comments in the code anyway...

Reply:
After reading @Sushisource I think I understand what is going on. Rust is *always* going to provide a default value, otherwise the program wouldn't compile since they are not Optional. The neon library just do the standard conversion in the bridge, so as long as they are coming from Core, they are always going to have a default value. That explains why the visibility API may see missing values but we don't...

@mjameswh can you confirm that the only path for this protos into the sdk-typescript is via core? If that's the case we should not make them optional, because they are not, and let other consumers make their own choices...

Reply:
Oh, I see where you're coming from... It seems that you may be a bit confused about how things work in the `@temporalio/client` package vs how they work in `@temporal/worker` and `@temporalio/workflows`.

The rest of this PR is all about stuff happening on the Worker side and inside the Workflow context. However, in this specific GitHub comment, we are looking at file `packages/client/src/types.ts`. That is part of the _client_ API (eg. `new Client().workflow.start(...)`).

In TS, the client API isn't based on Core SDK (and just to make it clear: TS is the only SDK that uses different gRPC implementations for its client vs its worker). Instead, it uses a pure JS gRPC implementation. The main reason for this is that it allows using the TS client in contexts where native code would not be possible.

So absolutely no Rust implied here.

> I don't think we should talk about the visibility API in the SDK comments, the UI could show the missing fields as "pending", or add a comment pop-up. In any case, most users of the UI are not going to look at the comments in the code anyway...

Sorry, I should have gave you a better introduction regarding the visibility API.

I used a screenshot of our web UI only as a very quick demonstration of the fact that some of the properties returned by the visibility API are known to be unavailable in some cases. However, the visibility API is not only about our Web UI. It is an API meant to be used programmatically by Temporal app developers (and yes, people are indeed building useful and interesting stuff based on that API).

For example, one can do:

```
  // Find workflows executions that might have been affected by some prior bug
  for await (const workflowInfo of client.workflow.list({ query: 'WorkflowType="MySuperCoolWorkflow"' })) {
    if (workflowInfo.status === 'RUNNING' && workflowInfo.historyLength > 15_000) {
      console.log(`Need to fix workflow: ${workflowInfo.workflowId}/${workflowInfo.runId}`);
    }
  }
```

Now, that `list()` function returns an async iterator of `WorkflowExecutionInfo`, ie. the same interface as what would be returned by `workflowHandle.describe()`. Note that this interface is part of the `@temporalio/client` public API, and essentially meant to be used by users of the Client API. `@temporalio/worker` and others never use the `WorkflowExecutionInfo` interface, nor do they ever need to parse an instance of the `IWorkflowExecutionInfo` protobuf message.

So, a user iterating on `WorkflowExecutionInfo` objects returned by the `list()` function could very naturally expect that properties of that object contain actual values, especially if there is no doc and no language annotation indicating that that property might not match expectations. Take the snippet I gave previously. It would return absolutely no workflow. Does that seems easy to figure out for a regular Temporal developer?

Reply:
Thanks, this is a great explanation. I didn't know that the client API bypasses core, and also it was common for end users to write their own visibility api tools/scripts...This is much clearer now!

To summarize, for the client path protobufjs will give us a default value (zero) but it may be the wrong value for the visibility API because it is eventually consistent, and programmatic clients will break. Also, hasOwnProperty() on default values will not work, breaking JSON.stringify() and many other iterators...

So you are right, we need a comment explaining all this. 

Not sure what to do about declaring it optional in the SDK, because the problem is not that a value could be missing, protobufjs will take care of that, the problem is that the default value may be completely wrong... And we can't tell whether it was explicitly set or is a default because it is not declared with an "optional" field in the proto definition. And if we change the proto we will break most clients, and nobody else uses "optional" in protos APIs anyway...

On the other hand, declaring it optional in Typescript, and setting the default value explicitly, even if it is redundant, it could be much more clear for someone reading the code... Anyway, your choice...

Reply:
I feel like having unset fields is preferable than using zero values, and is consistent with the other SDK APIs and general best practices in JS/TS.
Please note that the TS interface for for `WorkflowExecutionInfo` defines every field as nullable:
```
                    /** WorkflowExecutionInfo historySizeBytes */
                    historySizeBytes?: (Long|null);
```

Reply:
> To summarize, for the client path protobufjs will give us a default value (zero) but it may be the wrong value for the visibility API because it is eventually consistent

As it is now, a Workflow's `historyLength`, `historySizeBytes` and a few other attributes are only get sent to Elastic Search when Workflow Execution closes (see [here](https://github.com/temporalio/temporal/blob/8f11ba122fbbef537b07ca6b1823f143c9f36a0b/common/persistence/visibility/store/elasticsearch/visibility_store.go#L1012-L1020) and [here](https://github.com/temporalio/temporal/blob/8f11ba122fbbef537b07ca6b1823f143c9f36a0b/common/persistence/visibility/store/elasticsearch/visibility_store.go#L234-L238)). These values are not simply "lagging behind" as one would often expect for eventual consistency, they are simply not stored at all.

Another important thing: it is impossible for a workflow execution to actually have an history length of 0, or a size of 0. From the moment a workflow execution gets created, it already contains at least two events.

So lets take the `historySizeBytes` field as an example. There is only two possibilities:
- `protobufEncodedWorkflowExecutionInfo.historySizeBytes !== 0`: that value is valid (and presumably, it can no longer change, but let's not write this out, it case that might not be true for some other visibility engines);
- `protobufEncodedWorkflowExecutionInfo.historySizeBytes === 0`: that can only mean that the visibility store doesn't have that search attribute (presumably because the workflow execution has not completed yet, but actually, that's not even important). From a user's perspective, that is an undefined value.

> Also, hasOwnProperty() on default values will not work, breaking JSON.stringify() and many other iterators...

That's not an issue if they are using the high level client API (ie. `Client.workflow.list(...)`), as we don't return the protobuf objects to the user (well, we do expose the `raw`, as an escape hatch if they need something they can't get otherwise; but there's zero guarantee regarding that field, so no worry here).

Reply:
@bergundy That's the interface of the protobuf `WorkflowExecutionInfo` class, but we also have an interface of the same name in `@temporalio/client`, which unfortunately define `historyLength` as a non-optional property.

Making it optional at this point would be a breaking change. My worry on that one is really about usages of `workflowHandle.describe()` (existing uses of `list()` are less of a concern here, since if they were using the `historyLength` field, then they were incorrect anyway, so forcing them to fix this is not a bad thing). Similarly, `historySizeBytes` shouldn't be optional on return of `workflowHandle.describe()`, though it should be on return of `list()`.

So I'd say the best thing would be to introduce a new `WorkflowExecutionSummary` interface, which extends `WorkflowExecutionInfo` but make some of its field optional, then make `list()` return an iterator of `WorkflowExecutionSummary` instead of `WorkflowExecutionInfo`. That would follow what was done in Schedules.

@bergundy Does that make sense to you?

Reply:
I think we can expect history length to always be present on `WorkflowExecutionInfo` so it's fine that it's not optional.
And I support this for any other field for which we have this level of certainty that it will always be present.

The fact that `list` returns an empty field is wrong so we'd need to at least set the default value there.

> So I'd say the best thing would be to introduce a new WorkflowExecutionSummary interface, which extends WorkflowExecutionInfo but make some of its field optional, then make list() return an iterator of WorkflowExecutionSummary instead of WorkflowExecutionInfo. That would follow what was done in Schedules.

This may be our best option, I just learned today that the parent execution isn't set on list result either, making this approach even more attractive.

Reply:
historySizeBytes behavior mimics historyLength, I was just using whatever it was decided for historyLength. @bergundy 
 Shall I also leave it as not optional? It probably makes sense just to be consistent...

And for adding WorkflowExecutionSummary, shall we do that with the PR that fixes the parent execution or here?

Reply:
It depends on our confidence level for when this attribute will be non optional.

I think it was added a few releases ago, and if that's the case, let's keep it optional.

Reply:
> And for adding WorkflowExecutionSummary, shall we do that with the PR that fixes the parent execution or here?

I would wait on that for a followup PR. Don't want to unnecessarily block this one.

(apologize for the late comments)

Reply:
Ok, historySizeBytes is now optional and has a comment about >1.20 server release needed. There is also a system test using describe()